cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/AbstractArtifactLocalizer.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright © 2022 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.worker.sidecar;
+
+import com.google.common.io.CharStreams;
+import com.google.common.net.HttpHeaders;
+import io.cdap.cdap.api.retry.RetryableException;
+import io.cdap.cdap.common.ArtifactNotFoundException;
+import io.cdap.cdap.common.ServiceUnavailableException;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.common.utils.FileUtils;
+import io.cdap.cdap.proto.id.ArtifactId;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.StandardCopyOption;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Base class for Artifact Localizer and Artifact Cache.
+ * Provides functionality to fetch and cache artifacts from a remote endpoint.
+ */
+public abstract class AbstractArtifactLocalizer {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractArtifactLocalizer.class);
+  private final String dataDir;
+  private final RetryStrategy retryStrategy;
+
+  public AbstractArtifactLocalizer(String dataDir, RetryStrategy retryStrategy) {
+    this.dataDir = dataDir;
+    this.retryStrategy = retryStrategy;
+  }
+
+  public abstract HttpURLConnection openConnection(ArtifactId artifactId, @Nullable String peer) throws Exception;
```

Comment:
Missing javadoc for abstract method. Also, I think this method should be `protected` instead of `public`.

Reply:
It's no longer an abstract method after refactoring, but I have added javadoc for methods in the base class